File: ios/Tests/FlowbizOnsiteTests/CoreTestSupport.swift
// Test doubles and harness for the Slice 4 core (`FlowbizCore` with every
// seam faked): the tests drive the serial scheduler inline and observe the
// wire through `FakeHttpSender` bodies.
import Foundation
@testable import FlowbizOnsite

/// Recorded `ReachabilityMonitor`; tests fire `callback` to simulate
/// network restoration.
final class FakeReachability: ReachabilityMonitor, @unchecked Sendable {
    var started = false
    var callback: (@Sendable () -> Void)?

    func start(onNetworkAvailable: @escaping @Sendable () -> Void) {
        started = true
        callback = onNetworkAvailable
    }

    func stop() {
        started = false
        callback = nil
    }
}

/// Mutable box for the injected timezone offset (minutes).
final class OffsetBox: @unchecked Sendable {
    var value: Int
    init(_ value: Int) { self.value = value }
}

/// A `FlowbizCore` with every dependency faked. The `FakeTaskScheduler`
/// executes inline, so `core.track(...)` runs the whole pipeline (including
/// the flush drain) synchronously on the test thread.
final class CoreHarness: @unchecked Sendable {

    let config: FlowbizConfig
    let store: FakeKeyValueStore
    let clock: FakeClock
    let sender: FakeHttpSender
    let scheduler: FakeTaskScheduler
    let reachability: FakeReachability
    let offset: OffsetBox
    let queue: EventQueue
    let core: FlowbizCore

    init(
        config: FlowbizConfig = FlowbizConfig(appId: "77777"),
        store: FakeKeyValueStore = FakeKeyValueStore(),
        clock: FakeClock = FakeClock()
    ) {
        let sender = FakeHttpSender()
        let scheduler = FakeTaskScheduler()
        let reachability = FakeReachability()
        let offset = OffsetBox(-180)
        let queue = EventQueue(fileURL: temporaryQueueFile())
        self.config = config
        self.store = store
        self.clock = clock
        self.sender = sender
        self.scheduler = scheduler
        self.reachability = reachability
        self.offset = offset
        self.queue = queue
        self.core = FlowbizCore(
            config: config,
            store: store,
            queueFactory: { queue },
            sender: sender,
            scheduler: scheduler,
            clock: clock,
            deviceContext: DeviceContext(
                language: "pt-BR",
                screen: { "1170x2532" },
                timezoneOffsetMinutes: { _ in offset.value }
            ),
            reachability: reachability
        )
    }

    /// Entries of every sent body, in send order (flush batches flattened).
    func sentEntries() throws -> [[String: Any]] {
        try sender.bodies.flatMap { body -> [[String: Any]] in
            guard let object = try JSONSerialization.jsonObject(with: Data(body.utf8)) as? [String: Any],
                  let data = object["data"] as? [[String: Any]] else {
                throw FixtureSupport.FixtureError("unparseable body: \(body)")
            }
            return data
        }
    }

    func lastEntry() throws -> [String: Any] {
        guard let last = try sentEntries().last else {
            throw FixtureSupport.FixtureError("no entries were sent")
        }
        return last
    }
}

/// UUID-shape check (8-4-4-4-12 hex, any case) matching the Android tests.
func isUUIDShaped(_ value: Any?) -> Bool {
    guard let string = value as? String else { return false }
    return UUID(uuidString: string) != nil
}

/// Convenience accessors mirroring the Android JSONObject helpers.
func object(_ entry: [String: Any], _ key: String) -> [String: Any] {
    entry[key] as? [String: Any] ?? [:]
}

File: android/sdk/src/test/kotlin/br/com/flowbiz/onsite/CoreTestSupport.kt
package br.com.flowbiz.onsite

import org.json.JSONObject
import java.io.File

/**
 * Test doubles and harness for the Slice 4 core ([FlowbizCore] with every
 * seam faked): the tests drive the serial scheduler inline and observe the
 * wire through [FakeHttpSender] bodies.
 */

/** Recorded [ReachabilityMonitor]; tests fire [callback] to simulate network restoration. */
internal class FakeReachability : ReachabilityMonitor {
    var started = false
    var callback: Runnable? = null

    override fun start(onNetworkAvailable: Runnable) {
        started = true
        callback = onNetworkAvailable
    }

    override fun stop() {
        started = false
        callback = null
    }
}

/** Fixed-value [DeviceContext] with a mutable timezone offset. */
internal class FakeDeviceContext(
    override val language: String = "pt-BR",
    override val screen: String = "1080x2400",
    var offsetMinutes: Int = -180,
) : DeviceContext {
    override fun timezoneOffsetMinutes(wallMillis: Long): Int = offsetMinutes
}

/**
 * A [FlowbizCore] with every dependency faked. The [FakeTaskScheduler]
 * executes inline, so `core.track(...)` runs the whole pipeline (including
 * the flush drain) synchronously on the test thread.
 */
internal class CoreHarness(
    queueDir: File,
    val config: FlowbizConfig = FlowbizConfig(appId = "77777", baseUri = "https://store.com"),
    val store: FakeKeyValueStore = FakeKeyValueStore(),
    val clock: FakeClock = FakeClock(),
    val sender: FakeHttpSender = FakeHttpSender(),
    val scheduler: FakeTaskScheduler = FakeTaskScheduler(),
    val device: FakeDeviceContext = FakeDeviceContext(),
    val reachability: FakeReachability = FakeReachability(),
) {
    val queue = EventQueue(File(queueDir, "queue.jsonl"))
    val core = FlowbizCore(
        config = config,
        store = store,
        queueFactory = { queue },
        sender = sender,
        scheduler = scheduler,
        clock = clock,
        deviceContext = device,
        reachability = reachability,
    )

    /** Entries of every sent body, in send order (flush batches flattened). */
    fun sentEntries(): List<JSONObject> = sender.bodies.flatMap { body ->
        val data = JSONObject(body).getJSONArray("data")
        (0 until data.length()).map { data.getJSONObject(it) }
    }

    fun lastEntry(): JSONObject = sentEntries().last()
}
